lib/vdsm/storage/blockVolume.py

```diff
@@ -164,6 +164,13 @@ def getParent(self):
         """
         Return parent volume UUID
         """
+        parent_tag = self.getParentTag()
+        parent_meta = self.getParentMeta()
+        if parent_tag != parent_meta:
+            self.log.warning(
+                "volUUID=%s parent in tag (%s) and metadata (%s) differ",
+                self.volUUID, parent_tag, parent_meta)
+            self.refreshVolume()
```

Comment:
I'm not sure if this is the correct solution either.

We are assuming that the parent tag is outdated in the lvm cache. However, in a live merge, an `imageSyncVolumeChain` operation is called, and in `syncVolumeChain`, the first thing it does, even before updating the metadata, is `getChain`. Then we do `getParent` a couple of times. In those `getParent` I already see discrepancies in the parent UUID between tag and metadata.

Removing the lvm cache or reloading the lvs does not make them equal, metadata PUUID has not been updated yet. And raising here without handling the exception gets us stuck in a loop, invoking `imageSyncVolumeChain` over and over again.

So I think I would keep it as is and return the tag, so that the caller proceeds with the current chain, and the metadata gets updated. We just need to change the `refreshVolume` call with `lvm.reload_lvs(self.sdUUID)`, where:
```python
def reload_lvs(vg_name):
    _lvminfo._reloadlvs(vg_name)
```

Reply:
Here, the logs of a case where the check was triggered in an `merge`, and reload_lvs actually removed the volume that was pointed by the tag (so we avoid the bug!), and we return the correct parent  UUID:
```
2022-07-21 16:30:21,392+0200 INFO  (jsonrpc/5) [api.virt] START merge(drive={'poolID': '9227a275-ef9f-4816-81a0-506cbd7029b5', 'volumeID': 'a98deab9-b064-4073-becf-fb3e339fdde1', 'imageID': 'dda007f2-9825-4a7b-8b2e-62f2734a127c', 'domainID': 'a2ad7a23-96c5-4c23-9007-c0d1892ea07d'}, baseVolUUID='4cf4d502-dfed-491a-960c-142ada884fdc', topVolUUID='3a34c1b2-70a5-4b25-bf66-b3f3f0e21c70', bandwidth='0', jobUUID='098a4ab1-ecd3-4983-941d-7251b107d25b') from=::ffff:192.168.122.96,38222, flow_id=5cbdcbf6-e2bf-49ed-8a73-71553c8a544b, vmId=fa636499-0421-42f6-b537-6d28aeac6abf (api:48)
...
2022-07-21 16:30:21,486+0200 INFO  (jsonrpc/5) [vdsm.api] START getVolumeInfo(sdUUID='a2ad7a23-96c5-4c23-9007-c0d1892ea07d', spUUID='9227a275-ef9f-4816-81a0-506cbd7029b5', imgUUID='dda007f2-9825-4a7b-8b2e-62f2734a127c', volUUID='3a34c1b2-70a5-4b25-bf66-b3f3f0e21c70') from=::ffff:192.168.122.96,38222, flow_id=5cbdcbf6-e2bf-49ed-8a73-71553c8a544b, task_id=2aaf1719-04e4-4c65-8c32-deb2c0289258 (api:48)
2022-07-21 16:30:21,487+0200 INFO  (jsonrpc/5) [storage.volumemanifest] Info request: sdUUID=a2ad7a23-96c5-4c23-9007-c0d1892ea07d imgUUID=dda007f2-9825-4a7b-8b2e-62f2734a127c volUUID = 3a34c1b2-70a5-4b25-bf66-b3f3f0e21c70  (volume:243)
2022-07-21 16:30:21,503+0200 WARN  (jsonrpc/5) [storage.volumemanifest] volUUID=3a34c1b2-70a5-4b25-bf66-b3f3f0e21c70 parent in tag (e5710d00-ed07-4ae6-bfbf-e1dd37a601e7) and metadata (4cf4d502-dfed-491a-960c-142ada884fdc) differ (blockVolume:172)
2022-07-21 16:30:21,565+0200 WARN  (jsonrpc/5) [storage.lvm] Removing stale lv: a2ad7a23-96c5-4c23-9007-c0d1892ea07d/baca0c0c-dec6-4f74-bd5a-796878426646 (lvm:784)
2022-07-21 16:30:21,565+0200 WARN  (jsonrpc/5) [storage.lvm] Removing stale lv: a2ad7a23-96c5-4c23-9007-c0d1892ea07d/e5710d00-ed07-4ae6-bfbf-e1dd37a601e7 (lvm:784)
2022-07-21 16:30:21,565+0200 DEBUG  (jsonrpc/5) [storage.volumemanifest] New parent tag after reload=4cf4d502-dfed-491a-960c-142ada884fdc (blockVolume:175)
```

In this other occurrence, after the `imageSyncVolumeChain` is finished, the parent tag does not change, but is not the same as the metadata either. Raising here would cause an issue:
```
2022-07-21 16:30:22,879+0200 INFO  (merge/098a4ab1) [vdsm.api] FINISH imageSyncVolumeChain return=None from=internal, task_id=8b793feb-67d1-4412-9bc5-54ad39937580 (api:54)
2022-07-21 16:30:22,880+0200 INFO  (merge/098a4ab1) [vdsm.api] START getVolumeInfo(sdUUID='a2ad7a23-96c5-4c23-9007-c0d1892ea07d', spUUID='9227a275-ef9f-4816-81a0-506cbd7029b5', imgUUID='dda007f2-9825-4a7b-8b2e-62f2734a127c', volUUID='a98deab9-b064-4073-becf-fb3e339fdde1') from=internal, task_id=33bf3389-091b-42a8-abc2-c116216d0a78 (api:48)
2022-07-21 16:30:22,880+0200 INFO  (merge/098a4ab1) [storage.volumemanifest] Info request: sdUUID=a2ad7a23-96c5-4c23-9007-c0d1892ea07d imgUUID=dda007f2-9825-4a7b-8b2e-62f2734a127c volUUID = a98deab9-b064-4073-becf-fb3e339fdde1  (volume:243)
2022-07-21 16:30:22,899+0200 WARN  (merge/098a4ab1) [storage.volumemanifest] volUUID=a98deab9-b064-4073-becf-fb3e339fdde1 parent in tag (3a34c1b2-70a5-4b25-bf66-b3f3f0e21c70) and metadata (4cf4d502-dfed-491a-960c-142ada884fdc) differ (blockVolume:172)
2022-07-21 16:30:22,963+0200 DEBUG  (merge/098a4ab1) [storage.volumemanifest] New parent tag after reload=3a34c1b2-70a5-4b25-bf66-b3f3f0e21c70 (blockVolume:175)
```